cardano-testnet/src/Testnet/Property/Assert.hs

```diff
@@ -47,13 +50,28 @@ fileJsonGrep fp f = do
   let jsons = mapMaybe (Aeson.decode @Value) lines
   return $ L.any f jsons
 
+assertByDeadlineIOCustom
+  :: (MonadTest m, MonadIO m, HasCallStack)
+  => String -> DTC.UTCTime -> IO Bool -> m ()
+assertByDeadlineIOCustom str deadline f = GHC.withFrozenCallStack $ do
+  success <- H.evalIO f
```

Comment:
I think you could fork here and wait for the completion or kill the thread using [`timeout`](https://hackage.haskell.org/package/base-4.18.0.0/docs/System-Timeout.html#v:timeout) for example. Is there a point in waiting the `f` exceeds the deadline?